Handle new `source(…)`

RobinMalfait opened the pull request:
This PR introduces a new `source(…)` argument and improves on the existing `@source`. The goal of this PR is to make the automatic source detection configurable, let's dig in.

By default, we will perform automatic source detection starting at the current working directory. Auto source detection will find plain text files (no binaries, images, ...) and will ignore git-ignored files.

If you want to start from a different directory, you can use the new `source(…)` next to the `@import "tailwindcss/utilities" layer(utilities) source(…)`.

E.g.:

```css
/* ./src/styles/index.css */
@import 'tailwindcss/utilities' layer(utilities) source('../../');
```

Most people won't split their source files, and will just use the simple `@import "tailwindcss";`, because of this reason, you can use `source(…)` on the import as well:

E.g.:

```css
/* ./src/styles/index.css */
@import 'tailwindcss' source('../../');
```

Sometimes, you want to rely on auto source detection, but also want to look in another directory for source files. In this case, yuo can use the `@source` directive:

```css
/* ./src/index.css */
@import 'tailwindcss';

/* Look for `blade.php` files in `../resources/views` */
@source '../resources/views/**/*.blade.php';
```

However, you don't need to specify the extension, instead you can just point the directory and all the same automatic source detection rules will apply.

```css
/* ./src/index.css */
@import 'tailwindcss';

@source '../resources/views';
```

If, for whatever reason, you want to disable the default source detection feature entirely, and only want to rely on very specific glob patterns you define, then you can disable it via `source(none)`.

```css
/* Completely disable the default auto source detection */
@import 'tailwindcss' source(none);

/* Only look at .blade.php files, nothing else  */
@source "../resources/views/**/*.blade.php";
```

Note: even with `source(none)`, if your `@source` points to a directory, then auto source detection will still be performed in that directory. If you don't want that, then you can simply add explicit files in the globs as seen in the previous example.

```css
/* Completely disable the default auto source detection */
@import 'tailwindcss' source(none);

/* Run auto source detection in `../resources/views` */
@source "../resources/views";
```

RobinMalfait committed:
handle `source(…)`

Forward it from `@media source(…)` to `@tailwind utilities source(…)`.
Next, handle `@tailwind utilities source(…)` and retrieve the path.

RobinMalfait committed:
add integration test for `source(…)`

This integration test itself might change over time as we work through
this PR.

RobinMalfait committed:
forward correct `source(…)` to the oxide scanner

RobinMalfait commented:
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/tailwindlabs/tailwindcss/14767?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#14767** <a href="https://app.graphite.dev/github/pr/tailwindlabs/tailwindcss/14767?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈
* **#14766** <a href="https://app.graphite.dev/github/pr/tailwindlabs/tailwindcss/14766?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#14765** <a href="https://app.graphite.dev/github/pr/tailwindlabs/tailwindcss/14765?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* `next`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @RobinMalfait and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="11px" height="11px"/> <b>Graphite</b></a>
  <!-- Current dependencies on/for this PR: -->

RobinMalfait committed:
update test

RobinMalfait committed:
fix turbo input paths

We used to list `core` instead of `oxide`, this should make sure that
building Oxide works if we make changes in `oxide/`.

RobinMalfait committed:
fix some clippy warnings

RobinMalfait committed:
promote sources to auto source detection

RobinMalfait committed:
refactor, remove a level of nesting

thecrypticace commented on crates/oxide/src/lib.rs:
```diff
@@ -219,6 +219,36 @@ impl Scanner {
+            for source in sources {
+                // If a glob ends with `**/*`, then we just want to register the base path as a new
+                // base.
+                if source.pattern.ends_with("**/*") {
+                    source.pattern = source.pattern.trim_end_matches("**/*").to_owned();
+                }
```
1. `trim_end_matches` can be done unconditionally — no need to check `ends_with` first. I'd imagine the compiler is smart enough to omit the extra heap allocation in optimized builds.
2. `trim_end_matches` removes every occurrence which I think is fine. But the presence of the `ends_with` makes me think we wanted to do `strip_suffix` instead.

thecrypticace commented on crates/oxide/src/lib.rs:
```diff
@@ -219,6 +219,36 @@ impl Scanner {
+
+                // Contains a file extension, e.g.: `foo.html`, therefore we don't want to
+                // detect sources here.
+                if folder_name.to_str().unwrap().contains(".") {
```
While this is likely to be correct _most_ of the time — folder names can definitely have dots in them. I think the correct check here requires using the filesystem.

RobinMalfait committed:
merge `globs` and `detectSources`

Now that we can promote `@source "../path/to/folder"` to enable auto
source detection, it means that we can just merge the globs together and
add the base folder as if it was `@source "../path/to/folder"`

philipp-spiess committed:
Remove @tailwind utility nodes with parameters

RobinMalfait committed:
validate existence `source(…)` base path

RobinMalfait committed:
provide base and pattern separately

RobinMalfait committed:
add `bexpand` for expanding glob expressions

RobinMalfait committed:
expand patterns in `GlobEntry`

RobinMalfait committed:
run prettier

RobinMalfait committed:
update test name

RobinMalfait committed:
refactor: use variable directly

RobinMalfait committed:
update `@tailwindcss/postcss` with new `source(…)` setup

RobinMalfait committed:
update `@tailwindcss/vite` with new `source(…)` setup

RobinMalfait committed:
cleanup `console.log`

RobinMalfait committed:
add todo for follow up PR

RobinMalfait committed:
add CLI watch mode tests for `@source` and `source(…)`

RobinMalfait committed:
migrate detect sources to sources

RobinMalfait committed:
resolves base path in tests

Ensure symlinks are handled because `/var/` points to `/private/var/`.

RobinMalfait committed:
update all instances of `new Scanner(…)`

This now makes use of just `sources` instead of auto content detection.

thecrypticace commented on crates/oxide/src/lib.rs:
```diff
@@ -231,7 +218,58 @@ impl Scanner {
+                    return vec![source.clone()];
+                };
+
+                expression
+                    .into_iter()
+                    .filter_map(|expanded_pattern| expanded_pattern.map(|x| x.into()).ok())
+                    .map(move |pattern| GlobEntry {
+                        base: source.base.clone(),
+                        pattern,
+                    })
+                    .collect::<Vec<_>>()
```
```suggestion
                expression
                    .into_iter()
                    .filter_map(Result::ok)
                    .map(move |pattern| GlobEntry {
                        base: source.base.clone(),
                        pattern: pattern.into(),
                    })
                    .collect::<Vec<_>>()
```

RobinMalfait committed:
Update crates/oxide/src/lib.rs

Co-authored-by: Jordan Pittman <jordan@cryptica.me>

thecrypticace commented on packages/@tailwindcss-vite/src/index.ts:
```diff
@@ -379,9 +384,26 @@ class Root {
+        }
+
+        // No root specified, use the module graph
+        if (this.compiler.root === null) {
+          return []
+        }
```
```suggestion
        if (this.compiler.root === null) {
          this.includeCandidatesFromModuleGraph = false

          return []
        }
```

thecrypticace commented on packages/@tailwindcss-vite/src/index.ts:
```diff
@@ -379,9 +384,26 @@ class Root {
+        }
+
+        // TODO: In a follow up PR we want this filter this against the module graph.
+        this.includeCandidatesFromModuleGraph = false
```
```suggestion
        this.includeCandidatesFromModuleGraph = true
```

RobinMalfait committed:
Update packages/@tailwindcss-vite/src/index.ts

Co-authored-by: Jordan Pittman <jordan@cryptica.me>

RobinMalfait committed:
Update packages/@tailwindcss-vite/src/index.ts

Co-authored-by: Jordan Pittman <jordan@cryptica.me>

RobinMalfait commented:
closed in favor of https://github.com/tailwindlabs/tailwindcss/pull/14820
